luarules/gadgets/unit_area_timed_damage.lua

```diff
@@ -16,225 +16,521 @@ if not gadgetHandler:IsSyncedCode() then
-				Spring.SpawnExplosion(timedArea.x, timedArea.y, timedArea.z, 0, 0, 0, explosion)
-			else
-				StopTimedArea(frameAreas, index)
+local function cancelDelayedAreas(maxGameFrame)
+	local enabled, capacity, sx, sy, sz, dx, dy, dz, radiusTest
+	for shieldUnitID, radiusSq in pairs(shieldUnits) do
+		enabled, capacity = spGetUnitShieldState(shieldUnitID)
+		enabled = enabled and not spGetUnitIsStunned(shieldUnitID)
+		if enabled and capacity > 8 then
+			sx, sy, sz = spGetUnitPosition(shieldUnitID)
+			-- We don't care about the correctness of a short-lived, single-purpose array.
+			-- So when suppressing, we just replace a suppressed area's xpos with `false`.
+			for frame, delayedAreas in pairs(delayQueue) do
+				if frame <= maxGameFrame then
+					for ii = 1, #delayedAreas, 5 do
+						if delayedAreas[ii] then
+							dx = delayedAreas[ii  ] - sx
+							dy = delayedAreas[ii+1] - sy
+							dz = delayedAreas[ii+2] - sz
+							radiusTest = delayedAreas[ii+3].area_radius + radiusSq
+							if dx*dx + dz*dz < radiusTest and dy*dy < radiusTest then
+								delayedAreas[ii] = false
+							end
+						end
+					end
+				end
 			end
 		end
 	end
 end
 
 ---------------------------------------------------------------------------------------------------------------
+---- Gadget call-ins
 
 function gadget:Initialize()
-	for wdid, _ in pairs(WeaponAreaDefs) do
-		Script.SetWatchExplosion(wdid, true)
+	if not next(weaponTriggerParams) and not next(destroyTriggerParams) then
+		Spring.Log(gadget:GetInfo().name, LOG.INFO, "No timed area defs found. Removing gadget.")
+		gadgetHandler:RemoveGadget(self)
+	end
+
+	for weaponDefID, _ in pairs(weaponTriggerParams) do
+		Script.SetWatchExplosion(weaponDefID, true)
+	end
+
+	time  = 1 + (math.floor(Spring.GetGameFrame() / frameResolution))
+	frame = 1 + (math.floor(Spring.GetGameFrame() / frameResolution) % groupCount)
+	ticks = 1 + (Spring.GetGameFrame() % frameResolution)
+
+	timedAreas = {}
+	delayQueue = {}
+	shieldUnits = {}
+
+	for ii = 1, groupCount do
+		timedAreas[ii] = {}
+	end
+
+	if shieldsDenyAreas then
+		for _, unitID in pairs(Spring.GetAllUnits()) do
+			local unitDefID = Spring.GetUnitDefID(unitID)
+			if unitDefID and shieldUnitParams[unitDefID] then
+				shieldUnits[unitID] = shieldUnitParams[unitDefID]
+			end
+		end
 	end
 end
 
-function gadget:Explosion(weaponDefID, px, py, pz, attackerID, projectileID)
-	local params = WeaponAreaDefs[weaponDefID]
-	StartTimedArea(px, py, pz, params)
+function gadget:GameFrame(gameFrame)
+	ticks = ticks + 1
+	if ticks > frameResolution then
+		ticks = 1
+		frame = frame == groupCount and 1 or frame + 1
+		time  = time + 1
+		updateTimedAreas()
+	end
+
+	if ticks == shieldFrame then
+		cancelDelayedAreas(gameFrame + frameResolution - 1)
+	end
+
+	if delayQueue[gameFrame] then
+		local queue = delayQueue[gameFrame]
+		for ii = 1, #queue, 5 do
+			if queue[ii] then
+				startTimedArea(queue[ii], queue[ii+1], queue[ii+2], queue[ii+3], queue[ii+4])
```

Comment:
This is a very odd way to pass parameters to a function, and prone to bugs. Why are parameters being stored in a queue, why are they not grouped together in a table?

Reply:
reduces table usage to O(1) from O(n), idk if worth the maintainability decrease

Reply:
Lua doesn't give a nice way to do this tbh but I got over it. It's just not worth doing. Will fix.